abi/realloc.go

```diff
@@ -0,0 +1,52 @@
+package abi
+
+import "unsafe"
+
+// realloc allocates or reallocates memory for Component Model calls across
+// the host-guest boundary.
+//
+// Note: the use of uintptr assumes 32-bit pointers, e.g. GOOS=wasm32 when compiled for WebAssembly.
+//
+//go:export cabi_realloc
+//go:wasmexport cabi_realloc
+func realloc(ptr unsafe.Pointer, size, align, newsize uintptr) unsafe.Pointer {
+	if newsize <= size {
+		return unsafe.Add(ptr, offset(uintptr(ptr), align))
+	}
+	newptr := alloc(newsize, align)
+	if size > 0 {
+		copy(unsafe.Slice((*byte)(newptr), newsize), unsafe.Slice((*byte)(ptr), size))
```

Comment:
Why not deallocating the old memory block after copying?

Reply:
That’s an interesting example. If you realloc with a different alignment than the original allocation, presumably you’re storing a different type, in which case the copy shouldn’t happen. What does the spec say here?

Reply:
How? :)

Reply:
While standard Go doesn't have a the equivalent of `free()`, since we're deferring to Go's runtime in both cases we should allow them to be garbage collected if possible.   Unfortunately, there's no easy way to null out all references to the old pointer -- we just have to trust the client doesn't hold on to it.

Reply:
@Mossaka does the spec confirm that once the memory is allocated with `realloc` and the pointer is passed to a guest func, the memory is owned by the guest?

@dgryski will using `unsafe.Pointer` in `go:wasmexport` function signatures be sufficient to signal to the compiler that the value passed in is a pointer that can be GCed if not used?

Reply:
For garbage collection, if there are no references to the pointer in the guest address space, then it will be collected.  As with other uses of realloc, we'll need to be careful to keep track of the pointer in the guest if we don't want it accidentally reclaimed.

Reply:
i didn't find in the CM spec where it says the memory allocated with `realloc` will be owned by the guest...